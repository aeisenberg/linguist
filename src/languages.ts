import { ExtensionJustify } from './utils';

// tslint:disable-next-line
const languageMap = require('language-map');
// tslint:disable-next-line
// const lang = require('language-classifier');

interface ExtensionsTypes {
  [key: string]: string;
}

export interface DetectorOptions {}

/**
 * detecte program language through file extension
 *
 * @export
 * @class LanguageDetector
 */
export class Languages {
  extensionMap: {
    [key: string]: string;
  } = {};

  /**
   * Creates an instance of Detector.
   * @param {DetectorOptions} [options]
   * @memberof Detector
   */
  constructor() {
    this.extensionMap = this.loadExtensionMap();
  }

  /**
   * load language before detecting
   *
   * @private
   * @returns
   * @memberof Detector
   */
  private loadExtensionMap = () => {
    const extensions: ExtensionsTypes = {};

    Object.keys(languageMap).forEach((language) => {
      const languageMode = languageMap[language];
      const languageExtensions = (languageMode && languageMode.extensions) || [];
      languageExtensions.forEach((extension: string) => {
        extensions[extension.toLowerCase()] = language.toLowerCase();
      });
    });

    return Object.assign({}, extensions, ExtensionJustify);
  };

  /**
   * return extension map
   *
   * @returns
   * @memberof Detector
   */
  public getExtensionMap() {
    return this.extensionMap;
  }
}
