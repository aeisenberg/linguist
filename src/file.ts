/**
 * detect file info
 */

import * as fs from 'fs-extra';
import * as Path from 'path';
// @ts-ignore
import slash from 'slash2';

import { Languages } from './languages';

const SIMPLE_LINE_REG = /\s{1,}\/\/\s{1,}/g;
const ONLY_COMMENT_SIMPLE_LINE_REG = /\/\//g;
const ONE_LINE_REG = /$(\s{1,})\*.*/g;

export interface LineInfo {
  total: number;
  code: number;
  comment: number;
}

export interface FileInfo {
  name: string;
  languages: string;
  size: number;
  lines: LineInfo;
}

const DefaultLine: LineInfo = {
  total: 0,
  code: 0,
  comment: 0,
};

const DefaultFileInfo: FileInfo = {
  name: '',
  languages: '',
  size: 0,
  lines: DefaultLine,
};

/**
 * collect language info of a file
 *
 * @export
 * @class LocFile
 */
export class LocFile {
  private path: string;

  private language = new Languages();

  /**
   * Creates an instance of LocFile.
   * @param {string} filePath
   * @memberof LocFile
   */
  constructor(filePath: string) {
    if (!fs.existsSync(filePath)) {
      throw new Error('Error in File: file now exits.');
    }
    this.path = slash(filePath);
  }

  /**
   * get file type through a path
   *
   * @private
   * @param {string} path
   * @returns {string}
   * @memberof LocFile
   */
  getType(path: string): string {
    const fileExtension = `.${path.split('.').pop()}`;
    return this.language.extensionMap[fileExtension] || '';
  }

  public filterData = (data: string): LineInfo => {
    const lines = data.split(/\n/);
    let commentLength = 0;
    let codeLength = lines.length;
    const total = codeLength;

    lines.forEach((line) => {
      // 多行注释
      if (/(\*)|(\*\/)/g.test(line) || ONE_LINE_REG.test(line)) {
        commentLength += 1;
        codeLength -= 1;
      }
      // 单行注释
      if (SIMPLE_LINE_REG.test(line)) {
        commentLength += 1;
        // 只有注释
        if (ONLY_COMMENT_SIMPLE_LINE_REG.test(line)) {
          codeLength -= 1;
        }
      }
      if (!line) {
        codeLength -= 1;
      }
    });

    return {
      ...DefaultLine,
      total,
      code: codeLength,
      comment: commentLength,
    };
  };

  /**
   * get file info when LocFile init
   *
   * @private
   * @returns {FileInfo}
   * @memberof LocFile
   */
  getFileInfo(data?: string): FileInfo {
    let newData = data;
    const info: FileInfo = Object.assign({}, DefaultFileInfo);
    const name = this.path.split(Path.sep).pop() || '';
    try {
      const stat = fs.statSync(this.path);
      if (!stat.isFile()) {
        return info;
      }
      newData = data || fs.readFileSync(this.path, 'utf-8');
      info.name = name;
      info.size = (stat && stat.size) || 0;
      info.languages = this.getType(this.path);
      if (newData) {
        info.lines = this.filterData(newData);
      }
    } catch (err) {
      throw new Error('read file failed.');
    }
    return info;
  }

  /**
   * return file path
   *
   * @returns {string}
   * @memberof LocFile
   */
  public getPath(): string {
    return this.path;
  }

  public getFileInfoByContent(name: string, data: string): FileInfo {
    const info: FileInfo = Object.assign({}, DefaultFileInfo);
    info.name = name;
    info.languages = this.getType(name);
    info.lines = this.filterData(data);
    return info;
  }
}
