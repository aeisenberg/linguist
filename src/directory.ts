import globby from 'globby';
import fs from 'fs';
import path from 'path';
// @ts-ignore
import slash from 'slash2';

import { LineInfo, LocFile } from './file';

const defaultInfo: LineInfo = {
  total: 0,
  code: 0,
  comment: 0,
};

/**
 * Collect info of a directory.
 *
 * @export
 * @class LocDir
 */
export class LocDir {
  private pattern: string;

  /**
   * Creates an instance of LocDir.
   * @param {string} pattern
   * @memberof LocDir
   */
  constructor(pattern: string) {
    this.pattern = pattern;
  }

  /**
   * load directory info.
   *
   * @private
   * @returns {{
   *     files: LangFile[],
   *     info: LineInfo,
   *     languages: {
   *       [key: string]: number;
   *     }
   *   }}
   * @memberof LocDir
   */
  loadInfo(): {
    files: string[];
    info: LineInfo;
    languages: {
      [key: string]: LineInfo & {
        sum: number;
      };
    };
  } {
    const paths = globby.sync('**', {
      cwd: this.pattern,
      ignore: [
        '**/*.map',
        '**/yarn**',
        '**/.github',
        '**/node_modules/**',
        '**/dist/**',
        '**/*.snap',
      ],
    });
    const files: string[] = [];
    const info: LineInfo = { ...defaultInfo };
    let languages: {
      [key: string]: LineInfo & {
        sum: number;
      };
    } = {};

    paths.forEach((pathItem) => {
      const fullPath = slash(path.join(this.pattern, pathItem));
      const stat = fs.statSync(fullPath);
      if (!pathItem || !fs.existsSync(fullPath) || stat.isDirectory()) {
        return;
      }
      const file = new LocFile(fullPath);
      const fileLineInfo = file.getFileInfo();
      const { lines } = fileLineInfo;
      info.total += lines.total;
      info.code += lines.code;
      info.comment += lines.comment;
      const language = { ...languages[fileLineInfo.languages] };
      language.code = lines.code + (language.code || 0);
      language.sum = (language.sum || 0) + 1;
      language.comment = lines.comment + (language.comment || 0);
      language.total = lines.total + (language.total || 0);
      languages = {
        ...languages,
        [fileLineInfo.languages]: language,
      };
      files.push(fullPath);
    });

    return {
      files,
      info,
      languages,
    };
  }

  /**
   * Return detect pattern of the directory.
   *
   * @returns {string}
   * @memberof LocDir
   */
  public getPattern(): string {
    return this.pattern;
  }
}
