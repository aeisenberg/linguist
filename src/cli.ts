#!/usr/bin/env node

import chalk from 'chalk';
import program from 'commander';
// @ts-ignore
import slash from 'slash2';
import path from 'path';

import { LocDir } from './directory';
import { LocFile, LineInfo } from './file';
import { getVersion } from './utils';

program
  .version(getVersion(), '-v')
  .command('file <path>')
  .description('detect a file')
  .action(async (pathPattern) => {
    const info = await (new LocFile(pathPattern).getFileInfo());
    // eslint-disable-next-line no-console
    console.log(
      chalk.cyan(`
      path: \t\t${pathPattern}
      language: \t${info.languages}
      total lines: \t${String(info.lines.total)}
      code lines: \t${String(info.lines.code)}
      comment lines: \t${String(info.lines.comment)}
    `),
    );
  });

const formatInfo = (
  info: LineInfo,
  languages: {
    [key: string]: LineInfo & {
      sum: number;
    };
  },
) => `
  \ttotal lines: \t${String(info.total)}
  \tcode lines: \t${String(info.code)}
  \tcomment lines: \t${String(info.comment)}
  \t--------------------${Object.keys(languages)
    .map((key) => {
      const languageInfo = languages[key];
      return `\n\t${key.padEnd(10)} \t sum:${String(languageInfo.sum)} \ttotal:${String(
        languageInfo.total,
      )}  \tcomment:${String(languageInfo.comment)}  \tcode:${String(languageInfo.code)}`;
    })
    .join('')}`;

program.arguments('<cmd> [env]').action(async (cmd) => {
  const { info, languages } = await (new LocDir(slash(path.join(process.cwd(), cmd))).loadInfo());
  // eslint-disable-next-line no-console
  console.log(chalk.cyan(formatInfo(info, languages)));
});

program.parse(process.argv);
