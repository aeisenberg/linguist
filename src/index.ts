// @ts-ignore
import slash from 'slash2';
import fs from 'fs-extra';

import { LocDir } from './directory';
import { LocFile, LineInfo } from './file';

const loc = (
  fileOrDir: string,
): {
  info: LineInfo;
  files: string[];
  languages: {
    [key: string]: LineInfo & {
      sum: number;
    };
  };
} => {
  const stat = fs.statSync(slash(fileOrDir));
  if (stat.isFile()) {
    const locFile = new LocFile(slash(fileOrDir));
    const info = locFile.getFileInfo();
    const filePath = locFile.getPath();
    return {
      info: info.lines,
      files: [filePath],
      languages: { [info.languages]: { ...info.lines, sum: 1 } },
    };
  }
  const locDir = new LocDir(slash(fileOrDir));
  return locDir.loadInfo();
};

loc.locDir = LocDir;
loc.locFile = LocFile;

module.exports = loc;

export default loc;
