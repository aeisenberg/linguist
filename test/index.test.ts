/**
 *
 */
import * as path from 'path';
import slash from 'slash2';

import loc from '../src';
import { Languages } from '../src/languages';

describe('Languages', () => {
  beforeAll(() => {});
  describe('.getExtensionMap', () => {
    it('should return equals', () => {
      const map = new Languages().getExtensionMap();
      expect(map['.js']).toEqual('javascript');
      expect(map['.jsx']).toEqual('javascript');
      expect(map['.ts']).toEqual('typescript');
      expect(map['.tsx']).toEqual('typescript');
      expect(map['.cpp']).toEqual('c++');
      expect(map['.md']).toEqual('markdown');
      expect(map['.json']).toEqual('json');
      expect(map['.yml']).toEqual('yaml');
      expect(map['.svg']).toEqual('svg');
    });
  });
});

describe('LocFile', () => {
  it('js info', () => {
    const jsPath = slash(path.join(__dirname, '/data/index.js'));
    const file = loc(jsPath);
    expect(file.files[0]).toEqual(jsPath);
    const { info, languages } = file;
    expect(Object.keys(languages).pop()).toEqual('javascript');
    expect(info.total).toEqual(83);
    expect(info.code).toEqual(69);
    expect(info.comment).toEqual(6);
  });

  it('ts info', () => {
    const tsPath = slash(path.join(__dirname, '/data/index.ts'));
    const file = loc(tsPath);
    const { info, languages } = file;
    expect(Object.keys(languages).pop()).toEqual('typescript');
    expect(info.total).toEqual(169);
    expect(info.code).toEqual(98);
    expect(info.comment).toEqual(50);
  });

  it('dir info', () => {
    const tsPath = slash(path.join(__dirname, './'));
    const file = loc(tsPath);
    const { info, languages } = file;
    expect(Object.keys(languages).join()).toEqual('javascript,markdown,typescript');
    expect(info).toMatchSnapshot();
    expect(languages).toMatchSnapshot();
  });

  it('dir info', () => {
    const tsPath = slash(path.join(__dirname, '../example'));
    const file = loc(tsPath);
    const { info, languages } = file;
    expect(Object.keys(languages).join()).toEqual(
      'typescript,javascript,json,,svg,markdown,less,ejs',
    );
    expect(info).toMatchSnapshot();
    expect(languages).toMatchSnapshot();
  });
});
