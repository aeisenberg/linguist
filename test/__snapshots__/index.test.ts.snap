// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocFile dir info 1`] = `
Object {
  "code": 310,
  "comment": 68,
  "total": 445,
}
`;

exports[`LocFile dir info 2`] = `
Object {
  "javascript": Object {
    "code": 70,
    "comment": 6,
    "sum": 1,
    "total": 83,
  },
  "markdown": Object {
    "code": 83,
    "comment": 11,
    "sum": 1,
    "total": 126,
  },
  "typescript": Object {
    "code": 157,
    "comment": 51,
    "sum": 2,
    "total": 236,
  },
}
`;

exports[`LocFile dir info 3`] = `
Object {
  "code": 5223,
  "comment": 231,
  "total": 5950,
}
`;

exports[`LocFile dir info 4`] = `
Object {
  "": Object {
    "code": 0,
    "comment": 0,
    "sum": 6,
    "total": 0,
  },
  "ejs": Object {
    "code": 199,
    "comment": 0,
    "sum": 1,
    "total": 215,
  },
  "javascript": Object {
    "code": 45,
    "comment": 28,
    "sum": 2,
    "total": 81,
  },
  "json": Object {
    "code": 54,
    "comment": 4,
    "sum": 3,
    "total": 61,
  },
  "less": Object {
    "code": 489,
    "comment": 2,
    "sum": 12,
    "total": 558,
  },
  "markdown": Object {
    "code": 36,
    "comment": 0,
    "sum": 1,
    "total": 58,
  },
  "svg": Object {
    "code": 2,
    "comment": 0,
    "sum": 2,
    "total": 2,
  },
  "typescript": Object {
    "code": 4398,
    "comment": 197,
    "sum": 85,
    "total": 4975,
  },
}
`;
